### PR TITLE
Enable host group specification in a Cluster

### PR DESCRIPTION
**Pull request by EleanorRigby**
# Description

This PR enables an Admin to specify a host group in the vSphere clusters. 

This directly enables the Admin to define multiple AZs within the same cluster by using host groups. 

## Impacted Areas in Application
VM Placement & Creation

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Unit and Integration tests
# Checklist:

- [x] My code follows the standard ruby style guide
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_rule.rb [hunk 1]
```diff
@@ -32,6 +32,23 @@ def add_vm(vm)
       end
     end
 
+    # Adds VM to given VM Group & creates VM/HOST Affinity rule
+    # @param [Vim::VirtualMachine] vm
+    # @params [String] vm_group_name
```
Comment (ktchen14) [inline src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_rule.rb]: I think this should be `@param`.

File: src/vsphere_cpi/lib/cloud/vsphere/cluster_config.rb [hunk 2]
```diff
@@ -8,8 +8,8 @@ def initialize(name, config_hash)
       @config = config_hash
     end
 
-    def resource_pool
-      @config['resource_pool']
+    %w[vm_group host_group vm_host_affinity_rule_name resource_pool].each do |name|
```
Comment (ktchen14) [inline src/vsphere_cpi/lib/cloud/vsphere/cluster_config.rb]: This is terse but to be honest given how short this file is already is doing just:

```ruby
def vm_group; @config['vm_group']; end
def host_group; @config['host_group']; end
...
```

Is better in this case in my mind.

File: src/vsphere_cpi/lib/cloud/vsphere/drs_rules/vm_group.rb [hunk 3]
```diff
@@ -59,8 +59,37 @@ def delete_vm_groups(vm_group_names)
...
+        # Now, there is a potential problem here if anyone creates a rule
+        # outside of CPI using this vm_group, it will be deleted too. To narrow
+        # the scope of this potential bug, CPI will only delete the
+        # vm-host-group-affinity rule type. This is because, currently it only
+        # creates vm-host-affinity rule using vm groups.
+        #
+        # Long Term Solution is to disallow admin to specify their own rule
+        # names and CPI generates a rule name from vm_group and hence will be
+        # able to identify correct rule to be deleted.
+        rules_spec = empty_vm_groups.reduce([]) do |rules_spec, vm_group|
```
Comment (ktchen14) [inline src/vsphere_cpi/lib/cloud/vsphere/drs_rules/vm_group.rb]: In general `reduce` in Ruby is discouraged. I'm not exactly sure but I think in this case we can use `each_with_object` instead.

File: src/vsphere_cpi/lib/cloud/vsphere/resources/cluster.rb [hunk 4]
```diff
@@ -45,12 +46,72 @@ def free_memory
...
+      # Caches and returns the list of hosts under a cluster resource. It might
+      # be different from total hosts in a cluster if a host group is specified.
+      #
+      # In case a host group is present only the hosts under the host group are
+      # considered.
+      #
+      # If the host group does not exist/wrong it returns all the hosts
+      #
+      # @return [List] List of hosts in the cluster resource.
+      def host
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/resources/cluster.rb]: should be names hosts since it is returning a list of hosts and not a single host

File: src/vsphere_cpi/lib/cloud/vsphere/vcenter_client.rb [hunk 5]
```diff
@@ -458,6 +458,30 @@ def find_child_by_name(mob, child_path)
       find_child_by_name(mob.child_entity.find {|c| c.name == child_entity_name }, child_path)
     end
 
+    def clean_up_vm_groups(cluster, vm_group_names)
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/vcenter_client.rb]: Before merging should refactor this so that these methods are not duplicated in vm_group and here.

File: src/vsphere_cpi/lib/cloud/vsphere/resources/cluster.rb (hunk 4 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/lib/cloud/vsphere/resources/cluster.rb]: I agree. I will refactor this in another PR.  This was just kept host because of extensive usage and I did not want to change so many unrelated files.

File: src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_rule.rb [hunk 6]
```diff
@@ -59,6 +75,30 @@ def update_rule(rule_key)
       add_anti_affinity_rule(VimSdk::Vim::Option::ArrayUpdateSpec::Operation::EDIT, rule_key)
     end
 
+    def find_vm_group(vm_group_name)
```
Comment (EleanorRigby) [inline src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_rule.rb]: Don't think this is getting called from anywhere. Can delete it

File: src/vsphere_cpi/lib/cloud/vsphere/drs_rules/vm_group.rb [hunk 7]
```diff
@@ -59,15 +59,53 @@ def delete_vm_groups(vm_group_names)
           group_spec.remove_key = vm_group.name
           group_spec
         end
+
+        # If these vm_groups are getting deleted. Also delete all the
+        # vm-host group rules associated with it.
+        #
+        # @TODO : CAVEAT: if anyone creates a rule outside of CPI using these
+        # vm_groupS, IT will be deleted too. To narrow
```
Comment (EleanorRigby) [inline src/vsphere_cpi/lib/cloud/vsphere/drs_rules/vm_group.rb]: typo in comment

File: src/vsphere_cpi/lib/cloud/vsphere/drs_rules/vm_group.rb [hunk 8]
```diff
@@ -59,15 +59,53 @@ def delete_vm_groups(vm_group_names)
...
+        # the scope of this caveat, CPI will only delete the
+        # vm-host-group-affinity rule type. This is because, currently it only
+        # creates vm-host-affinity rule using vm groups.
+        #
+        # Long Term Solution is to disallow admin to specify their own rule
```
Comment (EleanorRigby) [inline src/vsphere_cpi/lib/cloud/vsphere/drs_rules/vm_group.rb]: remove this part of comment

File: src/vsphere_cpi/lib/cloud/vsphere/vcenter_client.rb (hunk 5 above)
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/vcenter_client.rb]: This method can be deleted as it is not being called from anywhere

File: src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb [hunk 9]
```diff
@@ -114,6 +114,36 @@ def create(vm_config)
           raise "Unable to parse vmx options: 'vmx_options' is not a Hash"
         end
 
+        host = nil
+        # Before cloning we need to make sure the host is correctly picked up
+        # from cluster's host group if provided.
+        unless cluster.host_group.nil?
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb]: this can be just if cluster.host_group

File: src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb [hunk 10]
```diff
@@ -114,6 +114,36 @@ def create(vm_config)
...
+          # placement_spec = VimSdk::Vim::Cluster::PlacementSpec.new
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb]: since, this Since, this is for future implementation, should remove it from here before we merge to master

File: src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb [hunk 11]
```diff
@@ -114,6 +114,36 @@ def create(vm_config)
...
+          #   logger.warning("received #{e} while asking DRS for recommendation")
+          #   host = cluster.host_group_mob.host.find do |host|
+          #     host.runtime.connection_state == 'connected' &&
+          #       !host.runtime.in_maintenance_mode
+          #   end
+          # end
+          # @TA:TODO : Stop-Gap measure until DRS team solves bug for placeVM below.
+          # Selecting a random host out of all host groups to get initial placement correct
+          # Filtering it for healthy host.
+          host = cluster.host_group_mob.host.find do |host|
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb]: if there is no healthy host, should you error out here?

File: src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb [hunk 12]
```diff
@@ -139,19 +170,27 @@ def create(vm_config)
...
 
           # DRS Rules
           create_drs_rules(vm_config, created_vm.mob, cluster)
 
-          # Add vm to VMGroup
-          add_vm_to_vm_group(vm_config, created_vm.mob, cluster)
+          # Add vm to VM Group present in vm_type
+          add_vm_to_vm_group(vm_config.vm_type.vm_group, created_vm_mob, cluster)
+
+          unless cluster.host_group.nil?
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb]: if instead of unless

File: src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb [hunk 13]
```diff
@@ -208,13 +247,24 @@ def create_drs_rules(vm_config, vm_mob, cluster)
...
+      return if vm_group.nil?
+      vm_group_creator = VSphereCloud::VmGroup.new(
+        @client,
+        cluster.mob
+      )
+      vm_group_creator.add_vm_to_vm_group(vm_mob, vm_group)
+    end
+
+    def create_vm_host_affinity_rule(vm_mob, cluster)
+      return if cluster.host_group.nil?
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb]: This check is already happening before the method is being called, either you can remove this, or remove the check when this method is called as, add_vm_to_vm_group method also has a check if vm_group.nil?

File: src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_lock.rb [hunk 14]
```diff
@@ -29,6 +29,8 @@ def with_drs_lock
 
     private
 
+    attr_reader :drs_lock_name
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_lock.rb]: Not needed, can be removed

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_lock_spec.rb [hunk 15]
```diff
@@ -2,7 +2,8 @@
 require 'timecop'
 
 describe VSphereCloud::DrsLock, fake_logger: true do
-  subject(:drs_lock) { described_class.new(vm_attribute_manager) }
+  subject(:drs_lock) { described_class.new(vm_attribute_manager, drs_lock_name) }
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_lock_spec.rb]: This is using the default lock_name so you dont need to pass it, in end there is a test to check a custom drs_lock name

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_lock_spec.rb [hunk 16]
```diff
@@ -2,7 +2,8 @@
...
+  let(:drs_lock_name) {'drs_lock'}
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_lock_spec.rb]: This can be removed refer comment before this

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_rule_spec.rb [hunk 17]
```diff
@@ -151,4 +151,48 @@ def self.it_adds_custom_attribute_to_vm
...
+      it 'returns the rule with DRS lock' do
+        expect(drs_lock).to receive(:with_drs_lock).and_yield.ordered
+        expect(datacenter_cluster).to_not receive(:reconfigure_ex)
+        expect(drs_rule).to_not receive(:create_vm_host_affinity_rule)
+        drs_rule.add_vm_host_affinity_rule(vm, vm_group_name, host_group_name)
+      end
+    end
+    context 'when rule does not exist' do
+      let(:rules) { [] }
+      it 'adds vm to the vm_group and creates DRS rule with DRS lock' do
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_rule_spec.rb]: reword, as this method only creates vm_host_afffinity_rule and does not add vm to vm_group

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_rule_spec.rb [hunk 18]
```diff
@@ -151,4 +151,48 @@ def self.it_adds_custom_attribute_to_vm
     end
   end
 
+  describe '#add_vm_host_affinity_rule' do
+    let(:vm) { instance_double('VimSdk::Vim::VirtualMachine', name: 'fake-vm') }
+    let(:fake_vm_1) { instance_double('VimSdk::Vim::VirtualMachine', name: 'fake-vm-1') }
+    let(:vm_group_name) {'fake-vm-group'}
+    let(:host_group_name) {'fake-host-group'}
+    let(:configuration_ex) { double(:configuration_ex, rule: rules, group: vm_groups) }
+    let(:vm_groups) { [ VimSdk::Vim::Cluster::VmGroup.new(name: vm_group_name, vm: [fake_vm_1]) ] }
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_rule_spec.rb]: this  cam be removed as this method doesn't use the vm_group

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_rule_spec.rb [hunk 19]
```diff
@@ -151,4 +151,48 @@ def self.it_adds_custom_attribute_to_vm
     end
   end
 
+  describe '#add_vm_host_affinity_rule' do
+    let(:vm) { instance_double('VimSdk::Vim::VirtualMachine', name: 'fake-vm') }
+    let(:fake_vm_1) { instance_double('VimSdk::Vim::VirtualMachine', name: 'fake-vm-1') }
+    let(:vm_group_name) {'fake-vm-group'}
+    let(:host_group_name) {'fake-host-group'}
+    let(:configuration_ex) { double(:configuration_ex, rule: rules, group: vm_groups) }
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_rule_spec.rb]: you can remove the group from here as it is not being used

File: src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_rule.rb [hunk 20]
```diff
@@ -32,6 +32,22 @@ def add_vm(vm)
       end
     end
 
+    # Adds VM to given VM Group & creates VM/HOST Affinity rule
+    # @param [Vim::VirtualMachine] vm
+    # @param [String] vm_group_name
+    # @param [String] host_group_name
+    def add_vm_host_affinity_rule(vm, vm_group_name, host_group_name)
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_rule.rb]: vm is not needed with current implementation. I think we can refactor later to make addition of vm to vm_group as part of this code rather then doing it separately

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_rule_spec.rb [hunk 21]
```diff
@@ -151,4 +151,48 @@ def self.it_adds_custom_attribute_to_vm
     end
   end
 
+  describe '#add_vm_host_affinity_rule' do
+    let(:vm) { instance_double('VimSdk::Vim::VirtualMachine', name: 'fake-vm') }
+    let(:fake_vm_1) { instance_double('VimSdk::Vim::VirtualMachine', name: 'fake-vm-1') }
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/drs_rule_spec.rb]: vm, fake_vm_1 can all be removed once the add_vm_host_affinity_rule definition is updated

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb [hunk 22]
```diff
@@ -107,4 +110,47 @@ def with_lock
...
     end
   end
+
+  describe '#get_delete_rule_spec' do
+    before do
+      allow(datacenter_cluster).to receive(:configuration_ex).and_return(configuration_ex)
+    end
+    context 'when no rule exists' do
+      before do
+        allow(configuration_ex).to receive(:rule).and_return(nil)
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb]: I think this returns an empty array, should confirm the with an actual call and setup the test accordingly

File: src/vsphere_cpi/lib/cloud/vsphere/drs_rules/vm_group.rb [hunk 23]
```diff
@@ -59,15 +59,53 @@ def delete_vm_groups(vm_group_names)
...
         end
       end
     end
 
     private
+    def get_delete_rule_spec(empty_vm_groups)
+      if @cluster.configuration_ex.rule.nil?
+        return nil
+      end
+      rules_spec= empty_vm_groups&.each_with_object([]) do |vm_group, rules_spec|
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/drs_rules/vm_group.rb]: dont need an & check as empty_vm_groups is an array and the check is being made when this is set

File: src/vsphere_cpi/lib/cloud/vsphere/drs_rules/vm_group.rb [hunk 24]
```diff
@@ -59,15 +59,53 @@ def delete_vm_groups(vm_group_names)
...
         @client.wait_for_task do
           @cluster.reconfigure_ex(config_spec, true)
         end
       end
     end
 
     private
+    def get_delete_rule_spec(empty_vm_groups)
+      if @cluster.configuration_ex.rule.nil?
+        return nil
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/drs_rules/vm_group.rb]: best would be this returns an empty array so return type is same and then called can check if spec is empty or not

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb [hunk 25]
```diff
@@ -107,4 +110,47 @@ def with_lock
...
+      end
+      it 'returns nil' do
+        expect(vm_group.send(:get_delete_rule_spec, *[vm_groups])).to be(nil)
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb]: as I commented should return empty array

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb [hunk 26]
```diff
@@ -107,4 +110,47 @@ def with_lock
...
+      end
+    end
+    context 'when vm_groups passed are empty' do
+      before do
+        allow(configuration_ex).to receive(:rule).and_return(['fake-rule'])
+      end
+      it 'returns nil' do
+        expect(vm_group.send(:get_delete_rule_spec, *[nil])).to be(nil)
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb]: spec says empty vm_groups but you are passing a nil here

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb [hunk 27]
```diff
@@ -107,4 +110,47 @@ def with_lock
...
+      end
+    end
+    context 'when vm groups passed are not empty' do
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb]: reword "when vm_groups are present"

File: src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb [hunk 28]
```diff
@@ -266,7 +266,8 @@ def fake_datastore(name, free_space, mob = nil)
             accessible_datastores: {
               'smaller-ds'=> smaller_ds,
             },
-            mob: cluster1_mob
+            mob: cluster1_mob,
+            host: [host_system],
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb]: then host can be removed from cluster1_mob and cluster2_mob definition as it is not being used

File: src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb [hunk 29]
```diff
@@ -262,6 +375,116 @@ def generate_host_property(mob:, name:, connection_state:, maintenance_mode:, me
...
+          it 'returns nil' do
+            expect(cluster.host_group).to be(nil)
+          end
+        end
+        context 'when host group is defined in config' do
+          let(:cluster_config) do
+            instance_double(
+              'VSphereCloud::ClusterConfig',
+              name: 'fake-cluster-name',
+              resource_pool: 'fake-resource-pool',
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb]: resource_pool can be removed

File: src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb [hunk 30]
```diff
@@ -262,6 +375,116 @@ def generate_host_property(mob:, name:, connection_state:, maintenance_mode:, me
...
+        it 'returns host_group name suffixed with _vm_group' do
+          expect(cluster.vm_group).to eq(cluster.host_group+VSphereCloud::Resources::Cluster::CLUSTER_VM_GROUP_SUFFIX)
+        end
+      end
+      context 'when host group name is longer than 100 characters' do
+        let(:cluster_config) do
+          instance_double(
+            'VSphereCloud::ClusterConfig',
+            name: 'fake-cluster-name',
+            resource_pool: 'fake-resource-pool',
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb]: resource_pool can be removed

File: src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb [hunk 31]
```diff
@@ -262,6 +375,116 @@ def generate_host_property(mob:, name:, connection_state:, maintenance_mode:, me
       end
     end
 
+    describe 'host groups and vm groups are defined in AZ' do
+      describe '#vm_group' do
+        let(:cluster_config) do
+          instance_double(
+            'VSphereCloud::ClusterConfig',
+            name: 'fake-cluster-name',
+            resource_pool: 'fake-resource-pool',
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb]: resource_pool can be removed

File: src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb [hunk 32]
```diff
@@ -262,6 +375,116 @@ def generate_host_property(mob:, name:, connection_state:, maintenance_mode:, me
...
+          it 'returns nil' do
+            expect(cluster.vm_host_affinity_rule_name).to be(nil)
+          end
+        end
+        context 'when host group is defined' do
+          let(:cluster_config) do
+            instance_double(
+              'VSphereCloud::ClusterConfig',
+              name: 'fake-cluster-name',
+              resource_pool: 'fake-resource-pool',
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb]: resource_pool can be removed and similarly below from all references

File: src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb [hunk 33]
```diff
@@ -262,6 +375,116 @@ def generate_host_property(mob:, name:, connection_state:, maintenance_mode:, me
...
+              host_group: 'fake-host-group',
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb]: you could just use let(:host_group) in all the tests instead of redefining cluster_config with host_group.

File: src/vsphere_cpi/spec/integration/vm_host_rule_spec.rb [hunk 34]
```diff
@@ -0,0 +1,402 @@
...
+        it 'should create one VM on the second cluster without host group\
+            (as resource pool has more memory available than host group)' do
+          simple_vm_lifecycle(cpi, '', vm_type, get_network_spec) do |vm_id|
+            vm = cpi.vm_provider.find(vm_id)
+            expect(vm.cluster).to eq(@second_cluster_name)
+          end
+          expect(get_count_vm_host_affinity_rules(cluster_mob) +
+            get_count_vm_host_affinity_rules(second_cluster_mob)).to eq(0)
+        end
+        it 'should create multiple VMs (3) one after another on the second cluster\'s resource pool' do
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/integration/vm_host_rule_spec.rb]: creation of multiple VM's for different configurations is repetitive and I think unnecessary for all the context in this spec file

File: src/vsphere_cpi/spec/integration/vm_host_rule_spec.rb [hunk 35]
```diff
@@ -0,0 +1,402 @@
...
+
+    context 'when there is only one vSphere cluster in an AZ' do
+      let(:hosts_in_first_host_group) { first_host_group_mob.host.map(&:name) }
+      let(:hosts_in_second_host_group) { second_host_group_mob.host.map(&:name) }
+
+      context 'when one host_group is specified in the cluster' do
+        it 'should create one VM on one of the hosts in the host groups' do
+          simple_vm_lifecycle(cpi, '', vm_type, get_network_spec) do |vm_id|
+            vm = cpi.vm_provider.find(vm_id)
+            expect(hosts_in_first_host_group).to include(vm.mob.runtime.host.name)
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/integration/vm_host_rule_spec.rb]: you should also test as part of this integration test that VM is added to the VM group and the rule is present. This spec is no where testing anything about VM Host Rule creation except that the count is 0 at the end

File: src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb (hunk 10 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb]: I would prefer keeping it. It is small and is not making things messy. Will be super easy to uncomment and push later.

Plus, this PR might uncomment it later depending on test runs with and without this change today.

File: src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_rule.rb (hunk 20 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_rule.rb]: `vm` is removed. I am not sure about merging addition of vm group to this function or I do not understand your intent. What  value would it bring rather than doing this way?

File: src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb (hunk 9 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb]: I would prefer explicit nil? everywhere as it differentiates `false` from `nil` .  makes it clear that host_group is not a boolean. It does not sees so from the name but practice has to be consistent across variables.

File: src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb (hunk 13 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb]: I had a TODO here before with title `should we make a function secure or the callee place should make sure of this` 

Since this is a local cluster resource and check will happen very fast, I decided to keep it here also in case if this function is used in future.
What do you say?

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb (hunk 22 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb]: added one more test to check empty. letting nil stay.

File: src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb (hunk 26 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/spec/unit/cloud/vsphere/drs_rules/vm_group_spec.rb]: Corrected.

File: src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb (hunk 33 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/spec/unit/cloud/vsphere/resources/cluster_spec.rb]: I did a bit of refactor. It would eliminate some redundancy but not much. i will keep it same for now.

File: src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb (hunk 28 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb]: It is being used. Caused error when i removed.

File: src/vsphere_cpi/spec/integration/vm_host_rule_spec.rb (hunk 34 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/spec/integration/vm_host_rule_spec.rb]: Letting this stay in favor of repeated testing for any non-deterministic bug.
Will take it our after the release is cut and feature stabilizes.

File: src/vsphere_cpi/spec/integration/vm_host_rule_spec.rb (hunk 35 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/spec/integration/vm_host_rule_spec.rb]: On it.

File: src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb (hunk 11 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/lib/cloud/vsphere/vm_creator.rb]: Done. Fixed. Good catch

File: src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_rule.rb (hunk 20 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/lib/cloud/vsphere/drs_rules/drs_rule.rb]: To be discussed later for future releases

File: src/vsphere_cpi/spec/integration/vm_host_rule_spec.rb (hunk 35 above)
Comment (nehagjain15) [reply src/vsphere_cpi/spec/integration/vm_host_rule_spec.rb]: Have created a follow up story to resolve this.
